protocol.go

```diff
@@ -297,7 +289,23 @@ func (p *Conn) readHeader() error {
 		}
 	}
 
-	header, err := Read(p.bufReader)
+	// For v1 the header length is at most 108 bytes.
+	// For v2 the header length is at most 52 bytes plus the length of the TLVs.
+	// We use 256 bytes to be safe.
+	const bufSize = 256
+	br := bufio.NewReaderSize(p.conn, bufSize)
+
+	header, err := Read(br)
+
+	if br.Buffered() != 0 {
+		buf := make([]byte, br.Buffered())
+		if _, err := br.Read(buf); err != nil {
+			return err // this should never as we read buffered data
+		}
```

Comment:
BTW err could be EOF

Reply:
> This is not correct as Read may read less than Buffered() so the proper way is to use io.ReadFull but then I am not sure it won't trigger another read from p.conn (i.e. I am not sure br.Buffered() will be 0 after that).

I think we can use Peek here to avoid allocation like:
```go
	if br.Buffered() > 0 {
		b, _ := br.Peek(br.Buffered())
		p.reader = io.MultiReader(bytes.NewReader(b), p.conn)
	} else {
		p.reader = p.conn
	}
```
Its not super documented but I think Peek's design is to avoid underlying Read if n <= Buffered()

Reply:
> Its not super documented but I think Peek's design is to avoid underlying Read if n <= Buffered()

Its not guaranteed unfortunately, see https://github.com/golang/go/issues/63548#issuecomment-1899223362

Reply:
@AlexanderYastrebov `bufio.Reader.Read` will never read here.

Call to read will jump directly [here](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/bufio/bufio.go;l=254), this is because:
* `n != 0` - len(p)
* `b.r == b.w` - there is something buffered

Also, note [definition of Buffered()](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/bufio/bufio.go;l=335).

Reply:
The question is about correctness.

There is simply no documented way to obtain buffered data from bufio.Reader and the current implementation can change in the future and this will break unexpectedly. See discussion https://github.com/golang/go/issues/63548 for more details.

From https://pkg.go.dev/bufio#Reader.Read
> To read exactly len(p) bytes, use io.ReadFull(b, p).

but that does not guarantee that reader does not buffer next chunk.

Therefore I think the right way is to capture everything that was read from the connection like I did in #119

Reply:
The probability they would change that Read behavior is close to 0.
But we sure can add a test that would guarantee that the Read does not touch the underlying reader.

At this point both branches are very similar, except for the fact that this one is easier to read and uses less resources.

I think you would agree this code is not very simple.

```
	bb := bytes.NewBuffer(make([]byte, 0, bufSize))
	br := bufio.NewReaderSize(io.TeeReader(p.conn, bb), bufSize)
```

Reply:
Can you, please, add such test?